src/main/java/it/smartcommunitylab/aac/webauthn/service/WebAuthnRpService.java

```
@@ -0,0 +1,269 @@
+        response.setOptions(options);
+        return response;
+    }
+
+    /**
+     * Returns:
+     * - the username of the authenticated user on successful authentication
+     * - null if the authentication was not successful
+     * - throws a WebAuthnAuthenticationException if some other error occourred
+     */
+    public String finishRegistration(
+            PublicKeyCredential<AuthenticatorAttestationResponse, ClientRegistrationExtensionOutputs> pkc,
+            String realm,
+            String key) throws WebAuthnAuthenticationException {
+        try {
+            final WebAuthnCredentialCreationInfo info = activeRegistrations.get(key);
+            WebAuthnAuthenticationException noSuchRegistration = new WebAuthnAuthenticationException("_",
+                    "Can not find matching active registration request");
+            if (info == null) {
+                throw noSuchRegistration;
+            }
+            if (!info.getRealm().equals(realm)) {
+                throw noSuchRegistration;
+            }
+            final String username = info.getUsername();
+            if (!StringUtils.hasText(username)) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Could not finish registration: missing username");
+            }
+            WebAuthnUserAccount existingAccount = webAuthnUserAccountService.findByProviderAndUsername(provider,
+                    username);
+            if (existingAccount != null) {
+                throw new WebAuthnAuthenticationException("_",
+                        "Account already registered");
+            }
+            final PublicKeyCredentialCreationOptions options = info.getOptions();
+            RegistrationResult result = rp
+                    .finishRegistration(FinishRegistrationOptions.builder().request(options).response(pkc).build());
+            boolean attestationIsTrusted = result.isAttestationTrusted();
+            if (attestationIsTrusted) {
```

if attestation is not trusted we should throw an exception

---

Addressed in cb95a78885936ee33c8f8d0d9ad6d34aed10c7bf